docs/lib/graphqlapi/fragments/android/authz/21_oidc.md

```diff
@@ -1,25 +1,118 @@
-Add the following code to your app:
+Implement your class which conforms to `OidcAuthProvider`:
+
+<amplify-block-switcher>
+<amplify-block name="Java">
+
+```java
+public final class MyOidcAuthProvider implements OidcAuthProvider {
+    @Override
+    public synchronized String getLatestAuthToken() throws ApiException {
+        final Semaphore semaphore = new Semaphore(0);
+        final AtomicReference<String> authToken = new AtomicReference<>();
+        final AtomicReference<AuthException> authError = new AtomicReference<>();
+        Amplify.Auth.fetchAuthSession(
+            session -> {
+                String token = ((AWSCognitoAuthSession) session)
+                        .getUserPoolTokens()
+                        .getValue()
+                        .getIdToken();
+                authToken.set(token);
+                semaphore.release();
+            },
+            error -> {
+                authError.set(error);
+                semaphore.release();
+            }
+        );
+
+        try {
+            semaphore.acquire();
+        } catch (Exception exception) {
+            throw new ApiException(
+                "Error fetching ID token.",
+                "Please try this operation again."
+            );
+        }
+        if (authError.get() != null) {
+            throw new ApiException(
+                "Error fetching ID token.",
+                authError.get(),
+                "Please try this operation again."
+            );
+        }
+        return authToken.get();
+    }
+}
```

Comment:
This is way, way, too crazy. I suggest we use RxAmplify:
```suggestion
OidcAuthProvider oidcAuthProvider = () ->
    RxAmplify.Auth.fetchAuthSession()
        .map(session -> ((AWSCognitoAuthSession) session)
            .getUserPoolTokens()
            .getValue()
            .getIdToken())
        .blockingGet();
```

Reply:
is our Rx library documented anywhere?

Reply:
rather than fully remove this, I added a separate tab for Rx.

Reply:
Yes, `fetchAuthSession()` has an Rx tab, [here](https://docs.amplify.aws/lib/auth/access_credentials/q/platform/android). The main docs site has an Rx landing page, [here](https://docs.amplify.aws/lib/project-setup/rxjava/q/platform/android). Here's an [AWS blog about it](https://aws.amazon.com/blogs/mobile/using-rxjava-with-aws-amplify-android-library/), and the [AWS launch notes](https://aws.amazon.com/about-aws/whats-new/2020/08/announcing-rxjava-support-amplify-android/).

Adding the RxJava tab is a good step in the right direction. However, I still think the Java and Kotlin examples are just too sprawling. You could use the `CompletableFuture` on the Android platform, to get a similar, shorter closure:

```kotlin
val authProviders = ApiAuthProviders.builder()
    .oidcAuthProvider {
        val future = CompletableFuture<String>()
        Amplify.Auth.fetchAuthSession(
            { future.complete((it as AWSCognitoAuthSession).userPoolTokens.value?.idToken) },
            { future.completeExceptionally(it) }
        )
        future.get()
    }
    .build()
Amplify.addPlugin(AWSApiPlugin(authProviders))
```

Reply:
addressed it with `CompletableFuture` in https://github.com/aws-amplify/docs/pull/2211/commits/d21779f557cd59ce71242d65938a97de30e2910d